File: CS/ViewSettingsSolution.Module/Controllers/ViewVariantsController.cs
using System;
using System.Collections.Generic;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Utils;
using DevExpress.ExpressApp.Model;
using ViewSettingsSolution.Module.BusinessObjects;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.Templates;

namespace ViewSettingsSolution.Module.Controllers {
    public class ViewVariantsController : ViewController {
        private SimpleAction DeleteViewVariantAction;
        private PopupWindowShowAction CreateViewVariantAction;
        private SingleChoiceAction SelectViewVariantAction;
        private SimpleAction UpdateCurrentViewVariantAction;
        private SimpleAction UpdateDefaultSettingsWithSelectedVariantAction;
        private string defaultUserSettings;
        private Boolean isLayoutProcessed = false;
        private Boolean isDefaultViewSelected = false;
        private ChoiceActionItem lastSelectedItem = null;

        private void SetDifferences(string xml, IModelView model) {
            Dictionary<string, string> differences = new Dictionary<string, string>();
            differences.Add("", xml);
            UserDifferencesHelper.SetUserDifferences(model, differences);
        }
        private Boolean TryLoadViewVariantFromXML(string xml) {
            Boolean result = false;
            View savedView = Frame.View;
            isLayoutProcessed = true;
            if(Frame.SetView(null, true, null, false)) {
                if(isDefaultViewSelected) {
                    UpdateDefaultSettings(savedView.Model);
                    isDefaultViewSelected = false;
                }
                SetDifferences(xml, savedView.Model);
                savedView.LoadModel(false);
                Frame.SetView(savedView);
                result = true;
            }
            isLayoutProcessed = false;
            return result;
        }
        private void SaveViewVariantToXML(SettingsStore store) {
            isLayoutProcessed = true;
            View.SaveModel();
            isLayoutProcessed = false;
            store.Xml = UserDifferencesHelper.GetUserDifferences(View.Model)[""];
            ((IObjectSpaceLink)store).ObjectSpace.CommitChanges();
        }
        private void UpdateDefaultSettings(IModelView model) {
            defaultUserSettings = UserDifferencesHelper.GetUserDifferences(model)[""];
        }
        private void UpdateActions(string itemToSelectCaption) {
            SelectViewVariantAction.Items.Clear();
            lastSelectedItem = null;
            CriteriaOperator criteria = CriteriaOperator.Parse("([ViewId] = ?) And ([IsShared] Or [OwnerId] is null Or [OwnerId] = ?)", View.Id, SecuritySystem.CurrentUserId);
            IObjectSpace objectSpace = Application.CreateObjectSpace(typeof(SettingsStore));
            foreach(SettingsStore item in objectSpace.GetObjects<SettingsStore>(criteria)) {
                SelectViewVariantAction.Items.Add(new ChoiceActionItem(item.Name, item));
            }
            if(SelectViewVariantAction.Items.Count > 0) {
                ChoiceActionItem defaultItem = new ChoiceActionItem("Default", null);
                SelectViewVariantAction.Items.Add(defaultItem);
                ChoiceActionItem itemToSelect = SelectViewVariantAction.Items.FindItemByID(itemToSelectCaption);
                SelectViewVariantAction.SelectedItem = (itemToSelect != null) ? itemToSelect : defaultItem;
                lastSelectedItem = SelectViewVariantAction.SelectedItem;
            }
            UpdateActionsActive();
        }
        private void UpdateActionsActive() {
            Boolean isActive = SelectViewVariantAction.Items.Count > 0 && SelectViewVariantAction.SelectedItem.Data != null;
            DeleteViewVariantAction.Active["HasVariants"] = isActive;
            UpdateCurrentViewVariantAction.Active["HasVariants"] = isActive;
            UpdateDefaultSettingsWithSelectedVariantAction.Active["HasVariants"] = isActive;
        }
        private void DeleteViewVariantAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            IObjectSpaceLink currentLayoutItem = SelectViewVariantAction.SelectedItem.Data as IObjectSpaceLink;
            if(TryLoadViewVariantFromXML(defaultUserSettings)) {
                IObjectSpace os = currentLayoutItem.ObjectSpace;
                os.Delete(currentLayoutItem);
                os.CommitChanges();
                isDefaultViewSelected = true;
                UpdateActions(null);
            }
        }
        private void CreateViewVariantAction_CustomizePopupWindowParams(object sender, CustomizePopupWindowParamsEventArgs e) {
            IObjectSpace objectSpace = e.Application.CreateObjectSpace(typeof(SettingsStore));
            e.View = e.Application.CreateDetailView(objectSpace, objectSpace.CreateObject<SettingsStore>());
        }
        private void CreateViewVariantAction_Execute(object sender, PopupWindowShowActionExecuteEventArgs e) {
            SettingsStore store = (SettingsStore)e.PopupWindowViewCurrentObject;
            if(SecuritySystem.CurrentUserId != null) {
                store.OwnerId = SecuritySystem.CurrentUserId.ToString();
            }
            store.ViewId = View.Id;
            SaveViewVariantToXML(store);
            string itemToSelectCaption = store.Name;
            isDefaultViewSelected = false;
            UpdateActions(itemToSelectCaption);
        }
        private void SelectViewVariantAction_Execute(object sender, SingleChoiceActionExecuteEventArgs e) {
            bool isVariantChanged = false;
            ChoiceActionItem currentItem = SelectViewVariantAction.SelectedItem;
            if(currentItem.Data != null) {
                if(TryLoadViewVariantFromXML(((SettingsStore)currentItem.Data).Xml)) {
                    isVariantChanged = true;
                }
            }
            else {
                if(TryLoadViewVariantFromXML(defaultUserSettings)) {
                    isDefaultViewSelected = true;
                    isVariantChanged = true;
                }
            }
            SelectViewVariantAction.SelectedItem = isVariantChanged ? currentItem : lastSelectedItem;
            UpdateActionsActive();
        }
        private void UpdateCurrentViewVariantAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            SaveViewVariantToXML(SelectViewVariantAction.SelectedItem.Data as SettingsStore);
        }
        private void UpdateDefaultSettingsWithSelectedVariantAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            SaveViewVariantToXML(SelectViewVariantAction.SelectedItem.Data as SettingsStore);
            UpdateDefaultSettings(View.Model);
        }
        private void View_ModelSaving(object sender, System.ComponentModel.CancelEventArgs e) {
            if(!isLayoutProcessed && !isDefaultViewSelected) {
                SetDifferences(defaultUserSettings, View.Model);
                e.Cancel = true;
            }
        }
        protected override void OnActivated() {
            base.OnActivated();
            View.ModelSaving += View_ModelSaving;
            if(!isLayoutProcessed) {
                UpdateDefaultSettings(View.Model);
                isDefaultViewSelected = true;
                UpdateActions(null);
            }
        }
        protected override void OnDeactivated() {
            View.ModelSaving -= View_ModelSaving;
            base.OnDeactivated();
        }
        public ViewVariantsController() {
            this.TargetViewNesting = Nesting.Root;
            this.DeleteViewVariantAction = new SimpleAction(this, "DeleteViewVariant", "Edit");
            this.DeleteViewVariantAction.Execute += new SimpleActionExecuteEventHandler(this.DeleteViewVariantAction_Execute);

            this.CreateViewVariantAction = new PopupWindowShowAction(this, "SaveAsNewViewVariant", "Edit");
            this.CreateViewVariantAction.CustomizePopupWindowParams += new CustomizePopupWindowParamsEventHandler(this.CreateViewVariantAction_CustomizePopupWindowParams);
            this.CreateViewVariantAction.Execute += new PopupWindowShowActionExecuteEventHandler(this.CreateViewVariantAction_Execute);

            this.SelectViewVariantAction = new SingleChoiceAction(this, "SelectViewVariant", "Edit");
            this.SelectViewVariantAction.PaintStyle = ActionItemPaintStyle.Caption;
            this.SelectViewVariantAction.Execute += new SingleChoiceActionExecuteEventHandler(this.SelectViewVariantAction_Execute);

            this.UpdateCurrentViewVariantAction = new SimpleAction(this, "UpdateCurrentViewVariant", "Edit");
            this.UpdateCurrentViewVariantAction.Execute += new SimpleActionExecuteEventHandler(this.UpdateCurrentViewVariantAction_Execute);

            this.UpdateDefaultSettingsWithSelectedVariantAction  = new SimpleAction(this, "UpdateDefaultViewVariant", "Edit");
            UpdateDefaultSettingsWithSelectedVariantAction.Execute += UpdateDefaultSettingsWithSelectedVariantAction_Execute;

            this.Actions.Add(this.DeleteViewVariantAction);
            this.Actions.Add(this.CreateViewVariantAction);
            this.Actions.Add(this.SelectViewVariantAction);
            this.Actions.Add(this.UpdateCurrentViewVariantAction);
            this.Actions.Add(this.UpdateDefaultSettingsWithSelectedVariantAction);
        }
    }
}

File: VB/ViewSettingsSolution.Module/Controllers/ViewVariantsController.vb
Imports System.Collections.Generic
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Utils
Imports DevExpress.ExpressApp.Model
Imports ViewSettingsSolution.Module.BusinessObjects
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.Templates

Namespace ViewSettingsSolution.Module.Controllers

    Public Class ViewVariantsController
        Inherits ViewController

        Private DeleteViewVariantAction As SimpleAction

        Private CreateViewVariantAction As PopupWindowShowAction

        Private SelectViewVariantAction As SingleChoiceAction

        Private UpdateCurrentViewVariantAction As SimpleAction

        Private UpdateDefaultSettingsWithSelectedVariantAction As SimpleAction

        Private defaultUserSettings As String

        Private isLayoutProcessed As Boolean = False

        Private isDefaultViewSelected As Boolean = False

        Private lastSelectedItem As ChoiceActionItem = Nothing

        Private Sub SetDifferences(ByVal xml As String, ByVal model As IModelView)
            Dim differences As Dictionary(Of String, String) = New Dictionary(Of String, String)()
            differences.Add("", xml)
            UserDifferencesHelper.SetUserDifferences(model, differences)
        End Sub

        Private Function TryLoadViewVariantFromXML(ByVal xml As String) As Boolean
            Dim result As Boolean = False
            Dim savedView As View = Frame.View
            isLayoutProcessed = True
            If Frame.SetView(Nothing, True, Nothing, False) Then
                If isDefaultViewSelected Then
                    UpdateDefaultSettings(savedView.Model)
                    isDefaultViewSelected = False
                End If

                SetDifferences(xml, savedView.Model)
                savedView.LoadModel(False)
                Frame.SetView(savedView)
                result = True
            End If

            isLayoutProcessed = False
            Return result
        End Function

        Private Sub SaveViewVariantToXML(ByVal store As SettingsStore)
            isLayoutProcessed = True
            View.SaveModel()
            isLayoutProcessed = False
            store.Xml = UserDifferencesHelper.GetUserDifferences(View.Model)("")
            CType(store, IObjectSpaceLink).ObjectSpace.CommitChanges()
        End Sub

        Private Sub UpdateDefaultSettings(ByVal model As IModelView)
            defaultUserSettings = UserDifferencesHelper.GetUserDifferences(model)("")
        End Sub

        Private Sub UpdateActions(ByVal itemToSelectCaption As String)
            SelectViewVariantAction.Items.Clear()
            lastSelectedItem = Nothing
            Dim criteria As CriteriaOperator = CriteriaOperator.Parse("([ViewId] = ?) And ([IsShared] Or [OwnerId] is null Or [OwnerId] = ?)", View.Id, SecuritySystem.CurrentUserId)
            Dim objectSpace As IObjectSpace = Application.CreateObjectSpace(GetType(SettingsStore))
            For Each item As SettingsStore In objectSpace.GetObjects(Of SettingsStore)(criteria)
                SelectViewVariantAction.Items.Add(New ChoiceActionItem(item.Name, item))
            Next

            If SelectViewVariantAction.Items.Count > 0 Then
                Dim defaultItem As ChoiceActionItem = New ChoiceActionItem("Default", Nothing)
                SelectViewVariantAction.Items.Add(defaultItem)
                Dim itemToSelect As ChoiceActionItem = SelectViewVariantAction.Items.FindItemByID(itemToSelectCaption)
                SelectViewVariantAction.SelectedItem = If(itemToSelect IsNot Nothing, itemToSelect, defaultItem)
                lastSelectedItem = SelectViewVariantAction.SelectedItem
            End If

            UpdateActionsActive()
        End Sub

        Private Sub UpdateActionsActive()
            Dim isActive As Boolean = SelectViewVariantAction.Items.Count > 0 AndAlso SelectViewVariantAction.SelectedItem.Data IsNot Nothing
            DeleteViewVariantAction.Active("HasVariants") = isActive
            UpdateCurrentViewVariantAction.Active("HasVariants") = isActive
            UpdateDefaultSettingsWithSelectedVariantAction.Active("HasVariants") = isActive
        End Sub

        Private Sub DeleteViewVariantAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            Dim currentLayoutItem As IObjectSpaceLink = TryCast(SelectViewVariantAction.SelectedItem.Data, IObjectSpaceLink)
            If TryLoadViewVariantFromXML(defaultUserSettings) Then
                Dim os As IObjectSpace = currentLayoutItem.ObjectSpace
                os.Delete(currentLayoutItem)
                os.CommitChanges()
                isDefaultViewSelected = True
                UpdateActions(Nothing)
            End If
        End Sub

        Private Sub CreateViewVariantAction_CustomizePopupWindowParams(ByVal sender As Object, ByVal e As CustomizePopupWindowParamsEventArgs)
            Dim objectSpace As IObjectSpace = e.Application.CreateObjectSpace(GetType(SettingsStore))
            e.View = e.Application.CreateDetailView(objectSpace, objectSpace.CreateObject(Of SettingsStore)())
        End Sub

        Private Sub CreateViewVariantAction_Execute(ByVal sender As Object, ByVal e As PopupWindowShowActionExecuteEventArgs)
            Dim store As SettingsStore = CType(e.PopupWindowViewCurrentObject, SettingsStore)
            If SecuritySystem.CurrentUserId IsNot Nothing Then
                store.OwnerId = SecuritySystem.CurrentUserId.ToString()
            End If

            store.ViewId = View.Id
            SaveViewVariantToXML(store)
            Dim itemToSelectCaption As String = store.Name
            isDefaultViewSelected = False
            UpdateActions(itemToSelectCaption)
        End Sub

        Private Sub SelectViewVariantAction_Execute(ByVal sender As Object, ByVal e As SingleChoiceActionExecuteEventArgs)
            Dim isVariantChanged As Boolean = False
            Dim currentItem As ChoiceActionItem = SelectViewVariantAction.SelectedItem
            If currentItem.Data IsNot Nothing Then
                If TryLoadViewVariantFromXML(CType(currentItem.Data, SettingsStore).Xml) Then
                    isVariantChanged = True
                End If
            Else
                If TryLoadViewVariantFromXML(defaultUserSettings) Then
                    isDefaultViewSelected = True
                    isVariantChanged = True
                End If
            End If

            SelectViewVariantAction.SelectedItem = If(isVariantChanged, currentItem, lastSelectedItem)
            UpdateActionsActive()
        End Sub

        Private Sub UpdateCurrentViewVariantAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            SaveViewVariantToXML(TryCast(SelectViewVariantAction.SelectedItem.Data, SettingsStore))
        End Sub

        Private Sub UpdateDefaultSettingsWithSelectedVariantAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            SaveViewVariantToXML(TryCast(SelectViewVariantAction.SelectedItem.Data, SettingsStore))
            UpdateDefaultSettings(View.Model)
        End Sub

        Private Sub View_ModelSaving(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs)
            If Not isLayoutProcessed AndAlso Not isDefaultViewSelected Then
                SetDifferences(defaultUserSettings, View.Model)
                e.Cancel = True
            End If
        End Sub

        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            AddHandler View.ModelSaving, AddressOf View_ModelSaving
            If Not isLayoutProcessed Then
                UpdateDefaultSettings(View.Model)
                isDefaultViewSelected = True
                UpdateActions(Nothing)
            End If
        End Sub

        Protected Overrides Sub OnDeactivated()
            RemoveHandler View.ModelSaving, AddressOf View_ModelSaving
            MyBase.OnDeactivated()
        End Sub

        Public Sub New()
            TargetViewNesting = Nesting.Root
            DeleteViewVariantAction = New SimpleAction(Me, "DeleteViewVariant", "Edit")
            AddHandler DeleteViewVariantAction.Execute, New SimpleActionExecuteEventHandler(AddressOf DeleteViewVariantAction_Execute)
            CreateViewVariantAction = New PopupWindowShowAction(Me, "SaveAsNewViewVariant", "Edit")
            AddHandler CreateViewVariantAction.CustomizePopupWindowParams, New CustomizePopupWindowParamsEventHandler(AddressOf CreateViewVariantAction_CustomizePopupWindowParams)
            AddHandler CreateViewVariantAction.Execute, New PopupWindowShowActionExecuteEventHandler(AddressOf CreateViewVariantAction_Execute)
            SelectViewVariantAction = New SingleChoiceAction(Me, "SelectViewVariant", "Edit")
            SelectViewVariantAction.PaintStyle = ActionItemPaintStyle.Caption
            AddHandler SelectViewVariantAction.Execute, New SingleChoiceActionExecuteEventHandler(AddressOf SelectViewVariantAction_Execute)
            UpdateCurrentViewVariantAction = New SimpleAction(Me, "UpdateCurrentViewVariant", "Edit")
            AddHandler UpdateCurrentViewVariantAction.Execute, New SimpleActionExecuteEventHandler(AddressOf UpdateCurrentViewVariantAction_Execute)
            UpdateDefaultSettingsWithSelectedVariantAction = New SimpleAction(Me, "UpdateDefaultViewVariant", "Edit")
            AddHandler UpdateDefaultSettingsWithSelectedVariantAction.Execute, AddressOf UpdateDefaultSettingsWithSelectedVariantAction_Execute
            Actions.Add(DeleteViewVariantAction)
            Actions.Add(CreateViewVariantAction)
            Actions.Add(SelectViewVariantAction)
            Actions.Add(UpdateCurrentViewVariantAction)
            Actions.Add(UpdateDefaultSettingsWithSelectedVariantAction)
        End Sub
    End Class
End Namespace
